server/batching/batcher.go

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2022-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package batching
+
+import (
+	"fmt"
+	"time"
+)
+
+type Context map[string]any
+
+const (
+	ContextBatchNumKey  = "batch_num"
+	ContextBatchSizeKey = "batch_size"
+)
+
+type Item func(ctx Context)
+type BatchCb func(ctx Context) error
+
+type Batcher struct {
+	cfg     Config
+	itemsCh chan Item
+	stopCh  chan struct{}
+	doneCh  chan struct{}
+	batches int
+}
+
+type Config struct {
+	// The frequency at which batches should be executed.
+	Interval time.Duration
+	// The maximum size of the queue of items.
+	Size int
```

Comment:
Well, channels must have a size, unless unbuffered which wouldn't work here as we need a queue. The practical use is to have a bound on resources we allocate. If we are unable to push because the channel is full it likely means we are under-provisioning and something is clearly wrong.